src/server/pfs/server/egress.go

```diff
@@ -0,0 +1,119 @@
+package server
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/miscutil"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/sdata"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset"
+	"github.com/pachyderm/pachyderm/v2/src/pfs"
+	"golang.org/x/net/context"
+)
+
+func getEgressPassword() (string, error) {
+	const passwordEnvar = "PACHYDERM_SQL_PASSWORD" // TODO move this to a package for sharing
+	password, ok := os.LookupEnv(passwordEnvar)
+	if !ok {
+		return "", errors.Errorf("must set %v", passwordEnvar)
+	}
+	return password, nil
+}
+
+func copyToObjectStorage(ctx context.Context, src Source, destURL string) (*pfs.EgressResponse_ObjectStorageResult, error) {
+	bytesWritten, err := getFileURL(ctx, destURL, src)
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	result := new(pfs.EgressResponse_ObjectStorageResult)
+	result.BytesWritten = bytesWritten
+	return result, nil
+}
+
+func copyToSQLDB(ctx context.Context, src Source, destURL string, fileFormat *pfs.SQLDatabaseEgress_FileFormat) (*pfs.EgressResponse_SQLDatabaseResult, error) {
+	url, err := pachsql.ParseURL(destURL)
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+
+	// do we need to wrap password in a secret?
+	password, err := getEgressPassword()
+	if err != nil {
+		return nil, err
+	}
+	db, err := pachsql.OpenURL(*url, password)
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	defer db.Close()
+
+	// all table are written through a single transaction
+	tx, err := db.BeginTxx(ctx, nil)
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	defer tx.Rollback()
+
+	// cache tableInfos because multiple files can belong to the same table
+	tableInfos := make(map[string]*pachsql.TableInfo)
+	result := new(pfs.EgressResponse_SQLDatabaseResult)
+	result.RowsWritten = make(map[string]int64)
+	err = src.Iterate(ctx, func(fi *pfs.FileInfo, file fileset.File) error {
+		if fi.FileType != pfs.FileType_FILE {
+			return nil
+		}
+
+		tableName := strings.Split(fi.File.Path, "/")[1]
+		tableInfo, ok := tableInfos[tableName]
+		if !ok {
+			// first time interacting with table, so do a full drop first
+			// TODO figure out how to full sync better
+			tableInfo, err = pachsql.GetTableInfoTx(tx, tableName)
+			if err != nil {
+				return errors.EnsureStack(err)
+			}
+			tableInfos[tableName] = tableInfo
+			result, err := tx.Exec(fmt.Sprintf("DELETE FROM %s.%s", tableInfo.Schema, tableInfo.Name))
+			if err != nil {
+				return errors.EnsureStack(err)
+			}
+			_, err = result.RowsAffected()
+			if err != nil {
+				return errors.EnsureStack(err)
+			}
```

Comment:
not really sure what this error would mean, but seems like the right way to handle it anyway

Reply:
I'm gonna just remove this since the some db drivers don't even implement `RowsAffected()`.